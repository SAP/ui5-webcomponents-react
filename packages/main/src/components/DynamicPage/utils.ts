export const DynamicPageCssVariables = {
  headerDisplay: '--_ui5wcr_DynamicPage_header_display',
  titleFontSize: '--_ui5wcr_DynamicPage_title_fontsize'
};
