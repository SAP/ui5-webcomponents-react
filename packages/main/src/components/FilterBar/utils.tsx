import { cloneElement } from 'react';

const inputTagNames = new Set([
  'UI5-COMBOBOX',
  'UI5-DATE-PICKER',
  'UI5-DATERANGE-PICKER',
  'UI5-DATETIME-PICKER',
  'UI5-DURATION-PICKER',
  'UI5-INPUT',
  'UI5-MULTI-INPUT',
  'UI5-RATING-INDICATOR',
  'UI5-STEP-INPUT',
  'UI5-TIME-PICKER'
]);

export const filterValue = (ref, child) => {
  const tagName = ref.tagName;
  let filterItemProps = {};
  if (inputTagNames.has(tagName)) {
    filterItemProps = { value: ref.value };
  }
  if (tagName === 'UI5-SELECT' || tagName === 'UI5-MULTI-COMBOBOX') {
    const selectedIndices = Array.from(ref.children as HTMLCollectionOf<any>)
      .map((item, index) => (item.selected ? index : false))
      .filter((el) => el !== false);
    const selectedIndicesSet = new Set(selectedIndices);
    const options = child.props.children.props.children?.map((item, index) => {
      if (selectedIndicesSet.has(index)) {
        return cloneElement(item, { selected: true });
      }
      return cloneElement(item, { selected: false });
    });
    filterItemProps = { children: options };
  }
  if (tagName === 'UI5-SWITCH' || tagName === 'UI5-CHECKBOX') {
    filterItemProps = { checked: ref.checked };
  }
  return filterItemProps;
};

export const renderSearchWithValue = (renderSearchElement, searchValue) => {
  return cloneElement(renderSearchElement, { value: searchValue ?? renderSearchElement?.props?.value });
};
