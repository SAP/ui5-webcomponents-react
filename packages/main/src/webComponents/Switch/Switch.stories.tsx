import type { Meta, StoryObj } from '@storybook/react';
import { Switch } from './index.js';

const meta = {
  title: 'Inputs / Switch',
  component: Switch,
  argTypes: {},
  args: {},
  tags: ['package:@ui5/webcomponents']
} satisfies Meta<typeof Switch>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Default: Story = {};
