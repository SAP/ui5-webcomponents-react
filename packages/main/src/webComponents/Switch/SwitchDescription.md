## Keyboard Handling

The state can be changed by pressing the Space and Enter keys.

## CSS Shadow Parts

<ui5-link target="_blank" href="https://developer.mozilla.org/en-US/docs/Web/CSS/::part">CSS Shadow Parts</ui5-link> allow developers to style elements inside the Shadow DOM.  
The `Switch` exposes the following CSS Shadow Parts:

*   slider - Used to style the track, where the handle is being slid
*   text-on - Used to style the `textOn` property text
*   text-off - Used to style the `textOff` property text
*   handle - Used to style the handle of the switch