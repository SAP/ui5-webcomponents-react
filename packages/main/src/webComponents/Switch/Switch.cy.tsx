import { cypressPassThroughTestsFactory } from '@/cypress/support/utils';
import { Switch } from './index';

describe('Switch', () => {
  it('Basic Test (generated)', () => {
    cy.mount(<Switch />);
    cy.get('[ui5-switch]').should('exist');
  });

  cypressPassThroughTestsFactory(Switch);
});
