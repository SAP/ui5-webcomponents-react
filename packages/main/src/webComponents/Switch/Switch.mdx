import { ControlsWithNote, DocsHeader, Footer } from '@sb/components';
import { Canvas, Meta } from '@storybook/blocks';
import * as ComponentStories from './Switch.stories';

<Meta of={ComponentStories} />

<DocsHeader since="0.4.0" />

<br />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

<Footer />
