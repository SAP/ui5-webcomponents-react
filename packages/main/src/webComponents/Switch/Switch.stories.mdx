import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs';
import { Switch } from '@ui5/webcomponents-react/dist/Switch';
import { CSSProperties, Ref } from 'react';
import { DocsHeader } from '@shared/stories/DocsHeader';

<Meta
  title="Inputs / Switch"
  component={Switch}
  args={{
    style: {},
    className: '',
    tooltip: '',
    slot: '',
    ref: null
  }}
/>

<DocsHeader />

<br />

## Example

<Canvas>
  <Story name="Default">
    {(args) => {
      return <Switch {...args} />;
    }}
  </Story>
</Canvas>

## Properties

<ArgsTable story="." />

## Keyboard Handling

The state can be changed by pressing the Space and Enter keys.
