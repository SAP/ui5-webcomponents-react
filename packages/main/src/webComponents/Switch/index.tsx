import { SwitchDesign } from '@ui5/webcomponents-react/dist/SwitchDesign';
import { withWebComponent } from '@ui5/webcomponents-react/dist/withWebComponent';
import { CommonProps } from '@ui5/webcomponents-react/interfaces/CommonProps';
import { Ui5CustomEvent } from '@ui5/webcomponents-react/interfaces/Ui5CustomEvent';

import '@ui5/webcomponents/dist/Switch';

export interface SwitchPropTypes extends Omit<CommonProps, 'onChange'> {
  /**
   * Defines if the component is checked.
   *
   * **Note:** The property can be changed with user interaction, either by cliking the component, or by pressing the `Enter` or `Space` key.
   */
  checked?: boolean;
  /**
   * Defines the component design.
   *
   * **Note:** If `Graphical` type is set, positive and negative icons will replace the `textOn` and `textOff`.
   */
  design?: SwitchDesign | keyof typeof SwitchDesign;
  /**
   * Defines whether the component is disabled.
   *
   * **Note:** A disabled component is noninteractive.
   */
  disabled?: boolean;
  /**
   * Defines the text, displayed when the component is not checked.
   *
   * **Note:** We recommend using short texts, up to 3 letters (larger texts would be cut off).
   */
  textOff?: string;
  /**
   * Defines the text, displayed when the component is checked.
   *
   * **Note:** We recommend using short texts, up to 3 letters (larger texts would be cut off).
   */
  textOn?: string;
  /**
   * Fired when the component checked state changes.
   */
  onChange?: (event: Ui5CustomEvent<HTMLInputElement>) => void;
}

/**
 * The `Switch` component is used for changing between binary states.
 * The component can display texts, that will be switched, based on the component state, via the `textOn` and `textOff` properties, but texts longer than 3 letters will be cutted off.
 * However, users are able to customize the width of `Switch` with pure CSS (`<Switch style="width: 200px">`), and set widths, depending on the texts they would use.
 * Note: the component would not automatically stretch to fit the whole text width.
 *
 * <ui5-link href="https://sap.github.io/ui5-webcomponents/playground/components/Switch" target="_blank">UI5 Web Components Playground</ui5-link>
 */
const Switch = withWebComponent<SwitchPropTypes>(
  'ui5-switch',
  ['design', 'textOff', 'textOn'],
  ['checked', 'disabled'],
  [],
  ['change']
);

Switch.displayName = 'Switch';

Switch.defaultProps = {
  checked: false,
  design: SwitchDesign.Textual,
  disabled: false
};

export { Switch };
