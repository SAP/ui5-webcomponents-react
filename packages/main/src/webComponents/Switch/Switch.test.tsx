import { render } from '@shared/tests';
import { createCustomPropsTest } from '@shared/tests/utils';
import { Switch } from '@ui5/webcomponents-react/dist/Switch';
import React from 'react';

describe('Switch', () => {
  test('Basic Test (generated)', () => {
    const { asFragment } = render(<Switch />);
    expect(asFragment()).toMatchSnapshot();
  });
  createCustomPropsTest(Switch);
});
