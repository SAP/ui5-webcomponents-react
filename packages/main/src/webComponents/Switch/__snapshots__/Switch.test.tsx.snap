// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch Basic Test (generated) 1`] = `
<DocumentFragment>
  <ui5-switch
    design="Textual"
    ui5-switch=""
  />
</DocumentFragment>
`;
